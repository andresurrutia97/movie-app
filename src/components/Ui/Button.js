import React from 'react';
import PropTypes from 'prop-types';

const Button = ({ children, onClick, type }) => {
  const baseClass = 'w-full focus:outline-none text-white py-2 px-4 rounded-md';

  let classes;

  switch (type) {
    case 'outlined':
      classes = [baseClass, 'border border-2 custom-blue-border bg-black'].join(
        ' '
      );
      break;

    default:
      classes = [baseClass, 'custom-blue-bg'].join(' ');
  }

  return (
    <button type="button" onClick={onClick} className={classes}>
      {children}
    </button>
  );
};

Button.propTypes = {
  children: PropTypes.string,
  onClick: PropTypes.func,
  type: PropTypes.string,
};

Button.defaultProps = {
  children: 'Click me',
  onClick: () => {},
  type: '',
};

export default Button;
