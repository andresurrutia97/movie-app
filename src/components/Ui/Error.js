import React from 'react';

const Error = () => (
  <p className="w-full py-5 text-center">Something went wrong</p>
);

export default Error;
