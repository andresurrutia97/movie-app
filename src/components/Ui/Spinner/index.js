import React from 'react';

import './index.css';

const Loading = () => <div className="spinner" />;

export default Loading;
