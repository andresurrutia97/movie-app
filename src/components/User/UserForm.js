import React from 'react';
import PropTypes from 'prop-types';

import Input from '../Ui/InputContainer';

const UserForm = ({ elements, onChange }) => {
  const form = elements.map((el) => {
    return (
      <Input
        key={el.id}
        title={el.config.elementName}
        elementtype={el.config.elementType}
        elementConfig={el.config.elementConfig}
        value={el.config.value}
        changed={(event) => onChange(event, el.id)}
        shouldValidate={el.config.validation}
        invalid={!el.config.valid}
        touched={el.config.touched}
      />
    );
  });
  return <div className="w-full"> {form}</div>;
};

UserForm.propTypes = {
  elements: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string,
      title: PropTypes.string,
      elementtype: PropTypes.string,
      elementConfig: PropTypes.array,
      value: PropTypes.string,
      changed: PropTypes.func,
      shouldValidate: PropTypes.bool,
      invalid: PropTypes.bool,
      touched: PropTypes.bool,
    })
  ),
  onChange: PropTypes.func,
};

UserForm.defaultProps = {
  elements: [],
  onChange: () => {},
};

export default UserForm;
