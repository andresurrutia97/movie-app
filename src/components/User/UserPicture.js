import React from 'react';
import PropTypes from 'prop-types';

import defaultUser from '../../img/default-user-image.png';

const UserPicture = ({ img }) => {
  return (
    <img
      className="h-32 w-32 mb-5 m-auto rounded-full shadow-lg shadow-md md:m-0 md:mr-12"
      src={img}
      alt="profile"
    />
  );
};

UserPicture.propTypes = {
  img: PropTypes.string,
};

UserPicture.defaultProps = {
  img: defaultUser,
};

export default UserPicture;
