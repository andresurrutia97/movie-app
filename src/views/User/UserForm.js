export default {
  name: {
    elementType: 'input',
    elementName: 'Name',
    elementConfig: {
      type: 'text',
      placeholder: 'Name',
    },
    value: '',
    validation: {
      required: true,
    },
    valid: false,
    touched: false,
    fullWidth: false,
  },
  tel: {
    elementType: 'input',
    elementName: 'Phone',
    elementConfig: {
      type: 'number',
      placeholder: 'Phone number',
    },
    value: '',
    validation: {
      required: true,
      isNumeric: true,
    },
    valid: false,
    touched: false,
    fullWidth: false,
  },
  img: {
    elementType: 'input',
    elementName: 'Picture',
    elementConfig: {
      type: 'text',
      placeholder: 'Picture url',
    },
    value: '',
    validation: {
      required: true,
    },
    valid: false,
    touched: false,
    fullWidth: false,
  },
  desc: {
    elementType: 'textarea',
    elementName: 'Description',
    elementConfig: {
      type: 'text',
      placeholder: 'Description',
    },
    value: '',
    validation: {
      required: true,
    },
    valid: false,
    touched: false,
    fullWidth: true,
  },
};
