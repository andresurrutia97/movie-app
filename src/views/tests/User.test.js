import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { Router, BrowserRouter } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import UserView from '../User';
import Spinner from '../../components/Ui/Spinner';
import Error from '../../components/Ui/Error';
import UserPicture from '../../components/User/UserPicture';
import UserForm from '../../components/User/UserForm';
import Button from '../../components/Ui/Button';

describe('<User/>', () => {
  let wrapper;
  const initProps = {
    userInfo: null,
    error: null,
    loading: false,
    userUpdate: {},
    loadingUpdate: false,
    errorUpdate: null,
    updateInfo: () => {},
  };
  const data = {
    data: {
      desc: 'Hi, im Andres and i like action movies and dogs',
      email: 'test@test.com',
      img: 'imgPath.jpg',
      name: 'Andres Urrutia Solarte f',
      tel: '3001231234',
    },
  };
  const mockSetState = jest.fn();
  const history = createMemoryHistory();
  const component = (
    <BrowserRouter>
      <UserView {...initProps} />
    </BrowserRouter>
  );

  beforeEach(() => {
    jest
      .spyOn(React, 'useState')
      .mockImplementation((init) => [init, mockSetState]);
    jest.spyOn(React, 'useCallback').mockImplementation((callback) => callback);
    jest.spyOn(React, 'useEffect').mockImplementationOnce((f) => f());
    jest.spyOn(history, 'goBack');
    wrapper = mount(
      <Router history={history}>
        <UserView {...initProps} />
      </Router>
    );
  });

  it('should render correcty ', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should contain a instance of <Spinner/> while no info ', () => {
    expect(wrapper.find(Spinner)).toHaveLength(1);
  });

  it('should contain an instance of <VideoPlayer/>, <InfoMovie/> if latestMovie prop contains info', () => {
    wrapper.setProps({
      children: <UserView userInfo={data} />,
    });
    expect(wrapper.find(UserForm)).toHaveLength(1);
    expect(wrapper.find(UserPicture)).toHaveLength(1);
  });

  it('should contain 1 instance of <Error/> if error prop caintains info ', () => {
    wrapper.setProps({ children: <UserView error={{ error: 'error' }} /> });
    expect(wrapper.find(Error)).toHaveLength(1);
  });

  it('should call the update user info function when click on the "save" button ', () => {
    const updateUserInfoMock = jest.fn();
    wrapper.setProps({
      children: <UserView userInfo={data} updateInfo={updateUserInfoMock} />,
    });
    wrapper.find(Button).at(1).simulate('click');
    expect(updateUserInfoMock).toBeCalledTimes(1);
  });

  it('should call the goBack() function of history when click on the "cancel" button ', () => {
    wrapper.setProps({
      children: <UserView userInfo={data} />,
    });
    wrapper.find(Button).at(0).simulate('click');
    expect(history.goBack).toBeCalledTimes(1);
  });
});
