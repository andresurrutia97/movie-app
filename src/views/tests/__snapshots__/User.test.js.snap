// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<User/> should render correcty  1`] = `
<div
  className="pt-16 my-4 mx-5 md:w-6/12 md:mx-auto md:mt-16"
>
  <div
    className="text-2xl text-white md:text-6xl text-center md:text-left"
  >
    Profile
  </div>
  <div
    className="spinner"
  />
</div>
`;
